### ISSUE TITLE
React-native synchronization

### ISSUE THREAD

**Issue by bootstraponline** [google/EarlGrey#518]
> Detox takes this into account by monitoring all asynchronous operations that take place in your app from the inside. Detox knows which network requests are currently in-flight. Detox knows how busy the React Native bridge is. Tests are automatically synchronized to the app and only move forward when the app is idle.
>
> https://github.com/wix/detox/blob/master/docs/Troubleshooting.Flakiness.md

It'd be nice to auto sync with React Native out of the box.

**Comment by LeoNatan** [comment]
[Maintainer of Detox iOS idling resources for RN]

I don’t see an issue in general to upstream our idling resources. Right now, there is a known limitation in some of our own idling resources, where we only support one bridge. For an upstream submission, these need to be ironed out.

Other than that, it’s a question of politics, if Google want to potentially dirty up Earl Grey with Facebook’s custom license.

Let’s see if there is demand to upstream our work.

**Comment by bootstraponline** [comment]
[react-native is BSD](https://github.com/facebook/react-native/blob/master/LICENSE) so I don't see an issue? What code from react-native is being used in the custom IdlingResource? If `NSClassFromString` is used then we might not need the Facebook code at all.

Hopefully other companies automating react-native with EarlGrey will comment on this issue.

**Comment by LeoNatan** [comment]
Are you sure it is not tainted by the custom FB license with controversial patents clause? (It does link and use React.js after all.)

At any rate, indeed Detox is indeed not linked directly with RN; it uses reflection to discover classes and use API, further isolating it from potentially being at odds with the FB license.

**Comment by bootstraponline** [comment]
> Are you sure it is not tainted by the custom FB license with controversial patents clause? (It does link and use React.js after all.)

No, you're confusing an additional patent grant with the actual license the code is released under.

> At any rate, indeed Detox is indeed not linked directly with RN; it uses reflection to discover classes and use API, further isolating it from potentially being at odds with the FB license.

That's perfect, there's no issue then.

**Comment by LeoNatan** [comment]
Can you invite me to the E.G. Slack?

**Comment by bootstraponline** [comment]
Yeah, what is your email? I'm code@bootstraponline.com

**Comment by bootstraponline** [comment]
PR has been opened to upstream the idle resource listeners:
https://github.com/google/EarlGrey/pull/703

**Comment by LeoNatan** [comment]
I am sorry, I will not getting involved in that PR, as per my explanation in the comment I left there. Please do with the PR as you will.
